crates/dbs-arch/src/aarch64/mod.rs

```diff
@@ -7,3 +7,58 @@
 pub mod gic;
 /// Logic for configuring aarch64 registers.
 pub mod regs;
+
+/// Trait for devices to be added to the Flattened Device Tree.
+pub trait DeviceInfoForFDT {
+    /// Returns the address where this device will be loaded.
+    fn addr(&self) -> u64;
+    /// Returns the associated interrupt for this device.
+    fn irq(&self) -> u32;
+    /// Returns the amount of memory that needs to be reserved for this device.
+    fn length(&self) -> u64;
+    /// Get device id
+    fn get_device_id(&self) -> Option<u32>;
+}
+
+/// MMIO device info used for FDT generating.
+#[derive(Debug, Clone, PartialEq)]
+pub struct MMIODeviceInfo {
+    /// MMIO address base
+    pub base: u64,
+    /// MMIO address size
+    pub size: u64,
+    /// Device irq
+    pub irqs: Vec<u32>,
+    /// Only virtio devices that support platform msi have device id
+    pub device_id: Option<u32>,
+}
+
+impl MMIODeviceInfo {
+    /// Create mmio device info.
+    pub fn new(base: u64, size: u64, irqs: Vec<u32>, device_id: Option<u32>) -> Self {
+        MMIODeviceInfo {
+            base,
+            size,
+            irqs,
+            device_id,
+        }
+    }
+}
+
+impl DeviceInfoForFDT for MMIODeviceInfo {
+    fn addr(&self) -> u64 {
+        self.base
+    }
+
+    fn irq(&self) -> u32 {
+        self.irqs[0]
```

Comment:
Is there any mechanism to ensure that `self.irqs[0]` is valid?
How should `MMIODeviceInfo::irqs` be organized?

Reply:
Currently only the business process guarantees that self.irqs[0] is valid.
1. Currently MMIODeviceInfo is only used for serial and virtio_mmio devices, which have one and only one irq.
2. MMIODeviceInfo will be created in the create_devices process, and the irqs[0] used in MMIODeviceInfo is in the subsequent system startup process, so the process ensures that self.irqs[0] is valid.
Based on the constraints of the use process in the above two points, is it necessary to add additional judgment conditions to ensure the validity of irqs[0]?

Reply:
It would be better to return an error instead of panic.
Oh, the function signature doesn't return error code. So maybe an assert to indicate the assumption?

Reply:
done, the return value of the irq() function has been modified to the result type.